File: core/src/androidTest/java/com/emarsys/core/connection/Connectivity_getConnectionState_ParameterizedTest.java
package com.emarsys.core.connection;


import static junit.framework.Assert.assertEquals;

import android.net.NetworkCapabilities;

import com.emarsys.core.concurrency.CoreSdkHandlerProvider;
import com.emarsys.testUtil.ConnectionTestUtils;
import com.emarsys.testUtil.InstrumentationRegistry;
import com.emarsys.testUtil.ReflectionTestUtils;
import com.emarsys.testUtil.TimeoutUtils;

import org.junit.Rule;
import org.junit.Test;
import org.junit.rules.TestRule;
import org.junit.runner.RunWith;
import org.junit.runners.Parameterized;

import java.util.Arrays;
import java.util.Collection;

@RunWith(Parameterized.class)
public class Connectivity_getConnectionState_ParameterizedTest {

    @Rule
    public TestRule timeout = TimeoutUtils.getTimeoutRule();

    @Parameterized.Parameter
    public boolean isConnected;

    @Parameterized.Parameter(1)
    public int connectionType;

    @Parameterized.Parameter(2)
    public ConnectionState connectionState;

    @Parameterized.Parameters
    public static Collection<Object[]> data() {
        return Arrays.asList(new Object[][]{
                {false, NetworkCapabilities.TRANSPORT_WIFI, ConnectionState.DISCONNECTED},
                {false, NetworkCapabilities.TRANSPORT_CELLULAR, ConnectionState.DISCONNECTED},
                {false, NetworkCapabilities.TRANSPORT_ETHERNET, ConnectionState.DISCONNECTED},
                {false, NetworkCapabilities.TRANSPORT_VPN, ConnectionState.DISCONNECTED},

                {true, NetworkCapabilities.TRANSPORT_WIFI, ConnectionState.CONNECTED},
                {true, NetworkCapabilities.TRANSPORT_CELLULAR, ConnectionState.CONNECTED_MOBILE_DATA},
                {true, NetworkCapabilities.TRANSPORT_ETHERNET, ConnectionState.CONNECTED},
                {true, NetworkCapabilities.TRANSPORT_VPN, ConnectionState.CONNECTED}
        });
    }

    @Test
    public void testGetConnectionState_connectionWatchDog() {
        ConnectionWatchDog connectionWatchDog = new ConnectionWatchDog(InstrumentationRegistry.getTargetContext(), new CoreSdkHandlerProvider().provideHandler());
        ReflectionTestUtils.setInstanceField(connectionWatchDog, "connectivityManager", ConnectionTestUtils.getConnectivityManagerMock(isConnected, connectionType));

        assertEquals(connectionState, connectionWatchDog.getConnectionState());
    }

}

File: core/src/androidTest/java/com/emarsys/core/connection/Connectivity_getConnectionState_ParameterizedTest.kt
package com.emarsys.core.connection

import android.net.NetworkCapabilities
import android.os.Build
import androidx.test.filters.SdkSuppress
import com.emarsys.core.concurrency.ConcurrentHandlerHolderFactory
import com.emarsys.testUtil.ConnectionTestUtils.getConnectivityManagerMock
import com.emarsys.testUtil.InstrumentationRegistry.Companion.getTargetContext
import com.emarsys.testUtil.ReflectionTestUtils.setInstanceField
import io.kotest.data.forAll
import io.kotest.data.row
import io.kotest.matchers.shouldBe
import kotlinx.coroutines.runBlocking
import org.junit.Test

@SdkSuppress(minSdkVersion = Build.VERSION_CODES.P)
class Connectivity_getConnectionState_ParameterizedTest  {

    @Test
    fun testGetConnectionState_connectionWatchDog() = runBlocking {
        forAll(
            row(false, NetworkCapabilities.TRANSPORT_WIFI, ConnectionState.DISCONNECTED),
            row(false, NetworkCapabilities.TRANSPORT_CELLULAR, ConnectionState.DISCONNECTED),
            row(false, NetworkCapabilities.TRANSPORT_ETHERNET, ConnectionState.DISCONNECTED),
            row(false, NetworkCapabilities.TRANSPORT_VPN, ConnectionState.DISCONNECTED),

            row(true, NetworkCapabilities.TRANSPORT_WIFI, ConnectionState.CONNECTED),
            row(
                true,
                NetworkCapabilities.TRANSPORT_CELLULAR,
                ConnectionState.CONNECTED_MOBILE_DATA
            ),
            row(true, NetworkCapabilities.TRANSPORT_ETHERNET, ConnectionState.CONNECTED),
            row(true, NetworkCapabilities.TRANSPORT_VPN, ConnectionState.CONNECTED)
        ) { isConnected, connectionType, expectedConnectionState ->
            val connectionWatchDog = ConnectionWatchDog(
                getTargetContext(), ConcurrentHandlerHolderFactory.create()
            )
            setInstanceField(
                connectionWatchDog,
                "connectivityManager",
                getConnectivityManagerMock(isConnected, connectionType)
            )

            connectionWatchDog.connectionState shouldBe expectedConnectionState
        }
    }
}